### PR TITLE
feat(*): template signing gsoc

### LINKED ISSUE #671: Method for template signing and verification in Template class
# Feature Request 🛍️

## Context

A method that will be used by template instances to sign the template markdown, logic and model.

## Detailed Description

Using crypto module from nodejs and node-forge library to execute this method. Use PKCS12 keystores to sign the contents. This method will be used in toArchive method. Users will pass the path to the PKCS12 keystore, password for the keystore as options to the toArchive method. From toArchive method the signTemplate method will be called. It will return the signature and it will be stored in the .cta archive.

### PR DESCRIPTION
**Pull request by sanketshevkar**
# Closes #671 

The developer while creating the archive can sign the template using a private key stored in a PKCS#12 keystore.

### Changes

- A new method added in the `Template` class that signs the hash of the contract text, logic and model of a template and returns an object containing signatory's metadata, timestamp, signatory's certificate, signature.
- A new node package added (node-forge), helps in handling keystores and crypto material inside it.
- An optional parameter, `keyStore` added to `toArchive` method in `Template` class using which the developer/author can pass the keystore path and password.
- An optional parameter `signatureObject` added to `toArchive` method of `TemplateSaver` class that will save the object containing signatory's metadata, timestamp, signatory's certificate, signature, in a `signature.json` file inside the .cta archive.
### Flags

- Test for the new and updated methods is yet to be written.

### Related Issues
- Issue #671 

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [x] Merging to `master` from `fork:branchname`

### CODE REVIEW

File: packages/cicero-cli/lib/commands.js [hunk 1]
```diff
@@ -512,6 +521,37 @@ class Commands {
...
+    /**
+     * Set default params before we verify signatures of template author/developer
+     *
+     * @param {object} argv the inbound argument values object
+     * @returns {object} a modfied argument object
+     */
+    static validateVerifyArgs(argv) {
+        argv = Commands.validateCommonArgs(argv);
+
+        if(!argv.target){
```
Comment (jeromesimeon) [inline packages/cicero-cli/lib/commands.js:533]: There is no `target` option for the `verify` CLI command, please remove this.

File: packages/cicero-cli/test/cli.js [hunk 2]
```diff
@@ -940,9 +940,33 @@ describe('#validateArchiveArgs', () => {
...
+        const newTemplate = await Template.fromArchive(fs.readFileSync(archiveName));
+        newTemplate.should.not.be.null;
+        newTemplate.should.have.own.property('authorSignature');
+        fs.unlinkSync(archiveName);
+    });
+
     it('should create a valid ergo archive', async () => {
         const archiveName = 'test.cta';
-        const result = await Commands.archive(template, 'ergo', archiveName);
+        const options = {
```
Comment (jeromesimeon) [inline packages/cicero-cli/test/cli.js:963]: We shouldn't need to pass any options if there is no keystone. This should work with `options = null` or `options = {}`. This would also mean that all existing tests should pass without changes.

This comment applies similarly to many of the other changes to CLI tests.

File: packages/cicero-cli/lib/commands.js [hunk 3]
```diff
@@ -496,7 +496,16 @@ class Commands {
...
+                let keystore = null;
+                if (options.keystore.path) {
+                    const p12File = fs.readFileSync(options.keystore.path, { encoding: 'base64' });
+                    const inputKeystore = {
+                        p12File: p12File,
+                        passphrase: options.keystore.passphrase
+                    };
+                    keystore = inputKeystore;
+                }
+                const archive = await template.toArchive(target, {keystore});
```
Comment (jeromesimeon) [inline packages/cicero-cli/lib/commands.js:508]: This means only the `keystore` option is being passed. Other options such as `warnings: true` should not be dropped.

File: packages/cicero-cli/test/cli.js [hunk 4]
```diff
@@ -1041,3 +1092,52 @@ describe('#get', async () => {
...
+            _: ['verify']
+        });
+        args.template.should.match(/cicero-cli[/\\]test[/\\]data[/\\]signedArchive$/);
+        args.target.should.match(/ergo/);
+    });
+    it('only target arg specified', () => {
+        process.chdir(path.resolve(__dirname, 'data/signedArchive/'));
+        const args  = Commands.validateVerifyArgs({
+            _: ['verify'],
+            target: 'ergo'
```
Comment (jeromesimeon) [inline packages/cicero-cli/test/cli.js:1108]: There is no `target` option for the `verify` command in the CLI. Those tests should be fixed accordingly.

File: packages/cicero-cli/test/cli.js [hunk 5]
```diff
@@ -1041,3 +1092,52 @@ describe('#get', async () => {
...
+    });
+    it('bad package.json', () => {
+        process.chdir(path.resolve(__dirname, 'data/'));
+        (() => Commands.validateVerifyArgs({
+            _: ['verify']
+        })).should.throw(' not a valid cicero template. Make sure that package.json exists and that it has a cicero entry.');
+    });
+});
+
+describe('#verify', async () => {
```
Comment (jeromesimeon) [inline packages/cicero-cli/test/cli.js:1136]: Please add a test where verification fails (so-called negative test).

File: packages/cicero-core/package.json [hunk 6]
```diff
@@ -84,11 +84,13 @@
     "@accordproject/markdown-slate": "0.14.0",
     "@accordproject/markdown-template": "0.14.0",
     "axios": "0.21.1",
+    "chai-fs": "^2.0.0",
```
Comment (jeromesimeon) [inline packages/cicero-core/package.json:87]: Why do we need this new dependency? Could we not use existing dependencies for this?

File: packages/cicero-core/src/template.js [hunk 7]
```diff
@@ -43,20 +44,26 @@ class Template {
...
 
     /**
      * Verifies that the template is well formed.
      * Compiles the Ergo logic.
      * Throws an exception with the details of any validation errors.
+     * @param {Object} options  - e.g., { verify: true }
      */
-    validate() {
+    validate(options = {}) {
+        if (options.verify) {
```
Comment (jeromesimeon) [inline packages/cicero-core/src/template.js:64]: Could you please rename this to `verifySignature` to avoid confusion.

File: packages/cicero-core/src/template.js [hunk 8]
```diff
@@ -172,14 +179,89 @@ class Template {
         return hasher.digest('hex');
     }
 
+    /**
+     * verifies the signature stored in the template object using the template hash and timestamp
+     * @return {boolean} true if signature is valid else false
```
Comment (jeromesimeon) [inline packages/cicero-core/src/template.js:184]: If the signature is invalid, this method throws an error. I would instead remove the `@return` parameter altogether.

File: packages/cicero-core/src/template.js [hunk 9]
```diff
@@ -172,14 +179,89 @@ class Template {
...
+        //convert public key in pem to public key type in node.
+        const publicKey = crypto.createPublicKey(publicKeyPem);
+        //signature verification process
+        const verify = crypto.createVerify('SHA256');
+        verify.write(templateHash + timestamp);
+        verify.end();
+        const result = verify.verify(publicKey, signature, 'hex');
+        if (!result) {
+            throw new Error('Template\'s author signature is invalid!');
+        }else{
```
Comment (jeromesimeon) [inline packages/cicero-core/src/template.js:207]: This can be removed (see previous comment) so that the method never returns a value.

File: packages/cicero-core/src/template.js [hunk 10]
```diff
@@ -172,14 +179,89 @@ class Template {
...
+        }
+    }
+
+    /**
+     * signs a string made up of template hash and time stamp using private key derived
+     * from the keystore
+     * @param {String} p12File - encoded string of p12 keystore file
+     * @param {String} passphrase - passphrase for the keystore file
+     * @param {Number} timestamp - timestamp of the moment of signature is done
+     * @private
```
Comment (jeromesimeon) [inline packages/cicero-core/src/template.js:218]: I think this method should be public? What do you think?

**Review by jeromesimeon — CHANGES_REQUESTED** [review]
This looks very nice. Much cleaner than the previous PR. I only have minor code style changes in addition to the following items:

- Could you document what's in the packages/cicero-cli/test/data/keystore.p12 file and how it was created? What credentials do they contain? The signed archives will have to be recreated whenever the templates change (which would happen whenever we create a new major version of Cicero).
- Something is not working properly with the GitHub CI workflow. I'm afraid you may have to create this PR again :(
- Also you should rebase as there is a conflict with `master`

File: packages/cicero-core/src/template.js (hunk 10 above)
Comment (sanketshevkar) [reply packages/cicero-core/src/template.js:218]: Sorry, my bad, it should be public I think.

File: packages/cicero-core/package.json (hunk 6 above)
Comment (sanketshevkar) [reply packages/cicero-core/package.json:87]: It's used to check if archive created using toArchive is a file and not empty [example](https://github.com/sanketshevkar/cicero/blob/99d4d09bbcfd8d71c16df1a67840c583f5cd7359/packages/cicero-core/test/template.js#L111). [Chai-fs](https://www.chaijs.com/plugins/chai-fs/) is an official fs plugin for chai library, so I thought it would be a good idea to use it. I can try the test cases using basic fs and chai methods if you want?

Comment (sanketshevkar) [reply packages/cicero-core/package.json]: I've have removed this dependency and simply checking, that the buffer created is not null. I did `npm uninstall --save chai-fs`
